[extension/storage/filestorage] Add CleanupOnStart flag for compaction temporary files

pandres-varian opened the pull request:
**Description:**
This PR includes a new flag **cleanup_on_start** for the compaction section.
During compaction a copy of the database is created, when the process is unexpectedly terminated that temporary file is not removed. That could lead to disk exhaustion given the following scenario: 
- Process is killed with a big database to be compacted
- Compaction is enabled on start
- Process will take longer to compact than the allotted time for the collector to reply health checks (see: #32456)
- Process is killed while compacting
- Big temporary file left

This mitigates the potential risk of those temporary files left in a short period of time, by this scenario or similar ones.

**Testing:**
Included corner case where two instances of the extensions are spawned and one is compacting while the other would attempt to cleanup.

**Documentation:** 
Included description in the README of the new configuration flag

Commit:
Add CleanupOnStart for compaction temporary files

Commit:
Merge branch 'main' into main

Commit:
Merge branch 'main' into main

Commit:
refactor: extract tempdb prefix to const

Commit:
fix: improve cleanup error logging

Commit:
Fix cleanup_on_start README

Co-authored-by: Daniel Jaglowski <jaglows3@gmail.com>

pandres-varian commented on extension/storage/filestorage/client.go:
```diff
@@ -342,3 +346,32 @@ func moveFileWithFallback(src string, dest string) error {
+
+// cleanup left compaction temporary files from previous killed process
+func (c *fileStorageClient) cleanup(compactionDirectory string) error {
+	pattern := filepath.Join(compactionDirectory, "tempdb*")
```
Thanks for your review!
Done in 602dc3a970421f5a6238e285b1471de006de7fcd

pandres-varian commented on extension/storage/filestorage/client.go:
```diff
@@ -342,3 +346,32 @@ func moveFileWithFallback(src string, dest string) error {
+		} else {
+			lockedCont++
+			c.logger.Debug("cleanup",
+				zap.String("lockedFile", item),
```
If I understood correctly, I have addressed this comment combined with the next one.

pandres-varian commented on extension/storage/filestorage/client.go:
```diff
@@ -342,3 +346,32 @@ func moveFileWithFallback(src string, dest string) error {
+	}
+	c.logger.Info("cleanup summary",
+		zap.Int("deletedFiles", delCont),
+		zap.Int("lockedFiles", lockedCont))
```
I'm not sure if this is what you meant: 820dcf9f98c0a6de46d42ec8631b88ff9910077f

Commit:
Merge branch 'main' into main

djaglowski commented on extension/storage/filestorage/client.go:
```diff
@@ -342,3 +347,30 @@ func moveFileWithFallback(src string, dest string) error {
+		return err
+	}
+
+	var errs error
```
```suggestion
	var errs []error
```

pandres-varian replied:
Thanks for the hint!

djaglowski commented on extension/storage/filestorage/client.go:
```diff
@@ -342,3 +347,30 @@ func moveFileWithFallback(src string, dest string) error {
+			c.logger.Debug("cleanup",
+				zap.String("deletedFile", item))
+		} else {
+			errs = errors.Join(errs, err)
```
```suggestion
			errs = append(errs, err)
```

djaglowski commented on extension/storage/filestorage/client.go:
```diff
@@ -342,3 +347,30 @@ func moveFileWithFallback(src string, dest string) error {
+	}
+	if errs != nil {
+		c.logger.Info("cleanup errors",
+			zap.Error(errs))
```
```suggestion
			zap.Error(errors.Join(errs...)))
```

djaglowski reviewed (approved):
Very minor nit, but use errors.Join once

Commit:
fix error logging aggregation

Co-authored-by: Daniel Jaglowski <jaglows3@gmail.com>

Commit:
Merge branch 'main' into main

Commit:
include change log file

Commit:
Merge branch 'main' into main

djaglowski commented on extension/storage/filestorage/client.go:
```diff
@@ -63,6 +65,9 @@ func newClient(logger *zap.Logger, filePath string, timeout time.Duration, compa
 	}
 
 	client := &fileStorageClient{logger: logger, db: db, compactionCfg: compactionCfg, openTimeout: timeout}
+	if compactionCfg.CleanupOnStart {
+		client.cleanup(compactionCfg.Directory)
+	}
```
One more thing - we should add a `Start` function so that cleanup is actually run on start, rather than on creation of the component.

Here's an example: https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/c6a6bd43875769d7400e8462e63730fe1a49ab11/extension/pprofextension/pprofextension.go#L31

pandres-varian replied:
Thanks for your patience, that hint was eye opening and could understand better the nuances of this extension lifecycle. 
It also address an issue according running the test on linux that was no simulating locked files. 
Please, take another look.

djaglowski replied:
Thanks for iterating on it.

Commit:
move clean_up from client to extension level

Each extension instance can create multiple clients.
Moving the logic to extension start is better aligned with the
extension lifecycle and prevent potential collisions, all clients will share the same compaction folder already cleaned up.

djaglowski commented on extension/storage/filestorage/extension.go:
```diff
@@ -42,6 +44,9 @@ func newLocalFileStorage(logger *zap.Logger, config *Config) (extension.Extensio
 // Start does nothing
 func (lfs *localFileStorage) Start(context.Context, component.Host) error {
+	if lfs.cfg.Compaction.CleanupOnStart {
+		lfs.cleanup(lfs.cfg.Compaction.Directory)
```
```suggestion
		return lfs.cleanup(lfs.cfg.Compaction.Directory)
```

djaglowski commented on extension/storage/filestorage/extension.go:
```diff
@@ -135,3 +140,30 @@ func isSafe(character rune) bool {
 }
+
+// cleanup left compaction temporary files from previous killed process
+func (c *localFileStorage) cleanup(compactionDirectory string) error {
```
```suggestion
func (lfs *localFileStorage) cleanup(compactionDirectory string) error {
```

djaglowski reviewed (commented):
Looks good, just some lint issues to address. You can run many of the checks locally with `make -C extension/storage/filestorage`

Commit:
fix lint errors

Commit:
Merge branch 'main' into main

Commit:
Merge branch 'main' into main
